File: Year-2023/LeetCode-July-Challenge/MaximumRunningTimeofNComputers.js
var maxRunTime = function (n, batteries) {
    let low = 1, high = 0;
    for (let x of batteries)
        high += x;

    const validate = (time, batteries, n) => {
        let total = 0;

        for (let x of batteries)
            total += Math.min(time, x);

        return total >= time * n;
    }

    let res = 0;

    while (low <= high) {
        let mid = low + Math.trunc((high - low) / 2);
        if (validate(mid, batteries, n)) {
            low = mid + 1;
            res = mid;
        } else high = mid - 1;
    }

    return res;
};

let n = 2, batteries = [3,3,3];
console.log(maxRunTime(n, batteries));

File: Year-2023/LeetCode-July-Challenge/MaximumRunningTimeofNComputers.java
class Solution {
    private boolean validate(long time, int[] batteries, int n) {
        long total = 0;
        for (int x : batteries)
            total += Math.min(time,  x);
        return total >= time * n;
    }

    public long maxRunTime(int n, int[] batteries) {
        long low = 1, high = 0;
        for (int x : batteries)
            high += x;

        long res = 0;
        while (low <= high) {
            long mid = low + (high - low) / 2;
            if (validate(mid, batteries, n)) {
                res = mid;
                low = mid + 1;
            } else
                high = mid - 1;
        }
        return res;
    }
}

public class MaximumRunningTimeofNComputers {
    public static void main(String[] args) {
        int n = 2;
        int[] batteries = { 3, 3, 3 };
        System.out.println(new Solution().maxRunTime(n, batteries));
    }
}